File: python/polypack/sync.py
"""Portable sync-envelope validation and checksum helpers."""

from __future__ import annotations

import json
import math
from pathlib import Path
from typing import Any, Iterable


def sync_checksum(operations: Iterable[dict[str, Any]]) -> str:
    """Return the TypeScript-compatible FNV-1a checksum for an ordered batch."""
    encoded = json.dumps(list(operations), ensure_ascii=False, separators=(",", ":"))
    digest = 2166136261
    units = encoded.encode("utf-16-le")
    for index in range(0, len(units), 2):
        digest ^= units[index] | (units[index + 1] << 8)
        digest = (digest * 16777619) & 0xFFFFFFFF
    return f"{digest:08x}"


def sync_identity_checksum(operation_ids: Iterable[str], transaction_ids: Iterable[str]) -> str:
    return sync_checksum([{
        "seq": 0,
        "timestamp": 0,
        "clientId": "sync-identities",
        "kind": "addNode",
        "payload": {"operationIds": sorted(operation_ids), "transactionIds": sorted(transaction_ids)},
    }])


def validate_sync_operation(operation: dict[str, Any]) -> None:
    if not isinstance(operation, dict):
        raise ValueError("sync operation must be an object")
    if not isinstance(operation.get("seq"), int) or isinstance(operation["seq"], bool) or operation["seq"] < 1:
        raise ValueError("sync sequence must be a positive integer")
    if not isinstance(operation.get("timestamp"), (int, float)) or isinstance(operation["timestamp"], bool) or not math.isfinite(operation["timestamp"]):
        raise ValueError("sync timestamp must be finite")
    if not isinstance(operation.get("clientId"), str) or not operation["clientId"]:
        raise ValueError("sync clientId must be non-empty")
    if not isinstance(operation.get("kind"), str) or not operation["kind"]:
        raise ValueError("sync operation kind must be non-empty")
    if not isinstance(operation.get("payload"), dict):
        raise ValueError("sync payload must be an object")
    for field in ("operationId", "transactionId"):
        if field in operation and (not isinstance(operation[field], str) or not operation[field]):
            raise ValueError(f"sync {field} must be non-empty")
    if "baseRevision" in operation and (not isinstance(operation["baseRevision"], int) or isinstance(operation["baseRevision"], bool) or operation["baseRevision"] < 0):
        raise ValueError("sync baseRevision must be a non-negative integer")


def validate_sync_batch(operations: Iterable[dict[str, Any]]) -> str:
    batch = list(operations)
    for operation in batch:
        validate_sync_operation(operation)
    return sync_checksum(batch)


class FileSyncOperationLog:
    """Small durable JSON operation log for the synchronous Python server."""

    def __init__(self, path: str | Path) -> None:
        self.path = Path(path)

    def load(self) -> dict[str, Any]:
        if not self.path.exists():
            return {"baseCursor": 0, "ops": [], "operationIds": [], "transactionIds": []}
        state = json.loads(self.path.read_text())
        if not isinstance(state, dict) or not isinstance(state.get("ops", []), list):
            raise ValueError("invalid sync operation log")
        return state

    def append_batch(self, operations: list[dict[str, Any]], base_cursor: int, operation_ids: set[str], transaction_ids: set[str]) -> None:
        state = {"baseCursor": base_cursor, "ops": operations, "operationIds": sorted(operation_ids), "transactionIds": sorted(transaction_ids)}
        state["checksum"] = sync_checksum(state["ops"])
        state["identityChecksum"] = sync_identity_checksum(state["operationIds"], state["transactionIds"])
        self.path.parent.mkdir(parents=True, exist_ok=True)
        temporary = self.path.with_suffix(self.path.suffix + ".tmp")
        temporary.write_text(json.dumps(state, separators=(",", ":")))
        temporary.replace(self.path)


class SyncServer:
    """Transport-neutral synchronous sync server matching the TS protocol."""

    def __init__(self, *, protocol_version: int = 1, max_ops: int | None = None, max_batch_ops: int | None = None,
                 authorize: Any = None, conflict: Any = None, operation_log: FileSyncOperationLog | None = None) -> None:
        if protocol_version < 1 or (max_ops is not None and max_ops < 1) or (max_batch_ops is not None and max_batch_ops < 1):
            raise ValueError("invalid sync server limits")
        self.protocol_version = protocol_version
        self.max_ops = max_ops
        self.max_batch_ops = max_batch_ops
        self.authorize = authorize
        self.conflict = conflict
        self.operation_log = operation_log
        self.base_cursor = 0
        self.ops: list[dict[str, Any]] = []
        self._operation_ids: set[str] = set()
        self._transaction_ids: set[str] = set()
        self._clients: dict[str, tuple[Any, Any, dict[str, Any]]] = {}
        if operation_log is not None:
            state = operation_log.load()
            self.base_cursor = int(state.get("baseCursor", 0))
            self.ops = list(state.get("ops", []))
            self._operation_ids = set(state.get("operationIds", []))
            self._transaction_ids = set(state.get("transactionIds", []))

    @property
    def cursor(self) -> int:
        return self.base_cursor + len(self.ops)

    def add_client(self, client_id: str, send: Any, filter: Any = None, metadata: dict[str, Any] | None = None) -> Any:
        self._clients[client_id] = (send, filter, metadata or {})
        return lambda message: self.receive(message, client_id)

    def remove_client(self, client_id: str) -> bool:
        return self._clients.pop(client_id, None) is not None

    def receive(self, message: dict[str, Any], sender_id: str | None = None) -> dict[str, Any]:
        if message.get("protocolVersion", self.protocol_version) != self.protocol_version:
            return self._send_ack(message, sender_id, [{"code": "protocol_version", "message": "unsupported sync protocol version"}])
        if message.get("type") == "request-snapshot":
            response = self.recover(int(message.get("fromSeq", 0)), self.max_batch_ops or len(self.ops) or 1, sender_id)
            self._send(sender_id, response)
            return response
        operations = list(message.get("ops", []))
        if self.max_batch_ops is not None and len(operations) > self.max_batch_ops:
            return self._send_ack(message, sender_id, [{"code": "batch_too_large", "message": "sync batch exceeds maxBatchOps"}])
        errors: list[dict[str, Any]] = []
        groups: dict[str, list[dict[str, Any]]] = {}
        for index, operation in enumerate(operations):
            validate_sync_operation(operation)
            groups.setdefault(f"tx:{operation.get('clientId')}:{operation.get('transactionId', index)}", []).append(operation)
        accepted: list[dict[str, Any]] = []
        for group in groups.values():
            group_errors: list[dict[str, Any]] = []
            for operation in group:
                context = {"clientId": sender_id or operation["clientId"], "protocolVersion": self.protocol_version}
                if self.authorize is not None and not self.authorize(operation, context):
                    group_errors.append({"code": "unauthorized", "message": "operation was not authorized", "operationId": operation.get("operationId")})
                if self.conflict is not None:
                    result = self.conflict(operation, context)
                    if result is False or isinstance(result, dict) and not result.get("ok"):
                        group_errors.append({"code": "conflict", "message": result.get("message", "operation conflicts") if isinstance(result, dict) else "operation conflicts", "operationId": operation.get("operationId")})
            if group_errors:
                errors.extend(group_errors)
            else:
                accepted.extend(group)
        self._commit(accepted)
        return self._send_ack(message, sender_id, errors)

    def recover(self, from_cursor: int, limit: int, client_id: str | None = None) -> dict[str, Any]:
        valid = self.base_cursor <= from_cursor <= self.cursor
        requested = from_cursor if valid else 0
        offset = requested - self.base_cursor if requested else 0
        page = self.ops[offset:offset + limit]
        visible = page
        if client_id in self._clients and self._clients[client_id][1] is not None:
            _, predicate, metadata = self._clients[client_id]
            context = {"clientId": client_id, "protocolVersion": self.protocol_version, "metadata": metadata}
            visible = [operation for operation in page if predicate(operation, context)]
        return {"type": "snapshot" if not requested else "delta", "clientId": "server", "fromSeq": requested,
                "cursor": (self.base_cursor if not requested else requested) + len(page), "more": offset + len(page) < len(self.ops),
                "ops": visible, "checksum": sync_checksum(visible), "protocolVersion": self.protocol_version,
                "errors": None if valid else [{"code": "cursor_expired", "message": "requested cursor is no longer available"}]}

    def _commit(self, operations: list[dict[str, Any]]) -> None:
        accepted: list[dict[str, Any]] = []
        accepted_transactions: set[str] = set()
        for operation in operations:
            operation_key = f"{operation['clientId']}:{operation['operationId']}" if operation.get("operationId") else None
            transaction_key = f"{operation['clientId']}:{operation['transactionId']}" if operation.get("transactionId") else None
            if operation_key in self._operation_ids or (transaction_key in self._transaction_ids and transaction_key not in accepted_transactions):
                continue
            if transaction_key:
                accepted_transactions.add(transaction_key)
            if operation_key:
                self._operation_ids.add(operation_key)
            if transaction_key:
                self._transaction_ids.add(transaction_key)
            accepted.append(operation)
        self.ops.extend(accepted)
        if self.max_ops is not None and len(self.ops) > self.max_ops:
            removed = len(self.ops) - self.max_ops
            self.ops = self.ops[removed:]
            self.base_cursor += removed
        if self.operation_log is not None:
            self.operation_log.append_batch(self.ops, self.base_cursor, self._operation_ids, self._transaction_ids)
        for operation in accepted:
            for client_id, (send, predicate, metadata) in self._clients.items():
                if client_id == operation.get("clientId"):
                    continue
                context = {"clientId": client_id, "protocolVersion": self.protocol_version, "metadata": metadata}
                visible = [operation] if predicate is None or predicate(operation, context) else []
                if visible:
                    send({"type": "delta", "clientId": "server", "fromSeq": self.cursor - len(accepted), "cursor": self.cursor, "ops": visible, "checksum": sync_checksum(visible), "protocolVersion": self.protocol_version})

    def _send(self, client_id: str | None, message: dict[str, Any]) -> None:
        if client_id in self._clients:
            self._clients[client_id][0](message)

    def _send_ack(self, message: dict[str, Any], client_id: str | None, errors: list[dict[str, Any]]) -> dict[str, Any]:
        response = {"type": "ack", "clientId": message.get("clientId", "server"), "fromSeq": message.get("fromSeq", 0), "ops": [], "protocolVersion": self.protocol_version, "errors": errors or None}
        self._send(client_id, response)
        return response

File: crates/polypack-core/src/sync.rs
//! Portable sync-envelope validation and checksums.
//!
//! This module deliberately stops at the wire contract. Transports, server
//! storage, authorization, and conflict resolution remain application-layer
//! concerns, matching the TypeScript sync implementation.

use crate::error::{PolypackError, Result};
use serde_json::{Map, Value};
use std::collections::HashSet;

/// Compute the TypeScript-compatible FNV-1a checksum for an ordered batch.
pub fn sync_checksum(operations: &[Value]) -> String {
    let input = serde_json::to_string(operations).expect("JSON sync values must serialize");
    let mut hash: u32 = 2_166_136_261;
    for unit in input.encode_utf16() {
        hash ^= unit as u32;
        hash = hash.wrapping_mul(16_777_619);
    }
    format!("{hash:08x}")
}

/// Compute the checksum used for retained operation/transaction identities.
pub fn sync_identity_checksum(operation_ids: &[String], transaction_ids: &[String]) -> String {
    let mut payload = Map::new();
    let mut operations = operation_ids.to_vec();
    let mut transactions = transaction_ids.to_vec();
    operations.sort();
    transactions.sort();
    payload.insert("operationIds".into(), Value::Array(operations.into_iter().map(Value::String).collect()));
    payload.insert("transactionIds".into(), Value::Array(transactions.into_iter().map(Value::String).collect()));
    let mut operation = Map::new();
    operation.insert("seq".into(), Value::from(0));
    operation.insert("timestamp".into(), Value::from(0));
    operation.insert("clientId".into(), Value::String("sync-identities".into()));
    operation.insert("kind".into(), Value::String("addNode".into()));
    operation.insert("payload".into(), Value::Object(payload));
    sync_checksum(&[Value::Object(operation)])
}

/// Validate the portable fields required by every sync operation.
pub fn validate_sync_operation(operation: &Value) -> Result<()> {
    let object = operation.as_object().ok_or_else(|| PolypackError::InvalidArgument("sync operation must be an object".into()))?;
    let seq = object.get("seq").and_then(Value::as_u64).ok_or_else(|| PolypackError::InvalidArgument("sync sequence must be a positive integer".into()))?;
    if seq == 0 { return Err(PolypackError::InvalidArgument("sync sequence must be a positive integer".into())); }
    if !object.get("timestamp").and_then(Value::as_f64).is_some_and(f64::is_finite) { return Err(PolypackError::InvalidArgument("sync timestamp must be finite".into())); }
    for field in ["clientId", "kind"] {
        if object.get(field).and_then(Value::as_str).is_none_or(str::is_empty) { return Err(PolypackError::InvalidArgument(format!("sync {field} must be non-empty"))); }
    }
    if !object.get("payload").is_some_and(Value::is_object) { return Err(PolypackError::InvalidArgument("sync payload must be an object".into())); }
    for field in ["operationId", "transactionId"] {
        if object.get(field).is_some() && object.get(field).and_then(Value::as_str).is_none_or(str::is_empty) { return Err(PolypackError::InvalidArgument(format!("sync {field} must be non-empty"))); }
    }
    if let Some(revision) = object.get("baseRevision") {
        if revision.as_u64().is_none() { return Err(PolypackError::InvalidArgument("sync baseRevision must be a non-negative integer".into())); }
    }
    Ok(())
}

/// Validate a batch and return its deterministic checksum.
pub fn validate_sync_batch(operations: &[Value]) -> Result<String> {
    for operation in operations { validate_sync_operation(operation)?; }
    Ok(sync_checksum(operations))
}

/// Synchronous, transport-neutral server state machine for native Rust hosts.
/// Durable storage and authorization are supplied by the host around this
/// state machine; the TypeScript and Python servers expose the same semantics.
pub struct SyncServer {
    protocol_version: u64,
    max_ops: Option<usize>,
    max_batch_ops: Option<usize>,
    base_cursor: u64,
    ops: Vec<Value>,
    operation_ids: HashSet<String>,
    transaction_ids: HashSet<String>,
    seen_ops: HashSet<String>,
}

impl SyncServer {
    pub fn new(protocol_version: u64, max_ops: Option<usize>, max_batch_ops: Option<usize>) -> Result<Self> {
        if protocol_version == 0 || max_ops == Some(0) || max_batch_ops == Some(0) { return Err(PolypackError::InvalidArgument("invalid sync server limits".into())); }
        Ok(Self { protocol_version, max_ops, max_batch_ops, base_cursor: 0, ops: Vec::new(), operation_ids: HashSet::new(), transaction_ids: HashSet::new(), seen_ops: HashSet::new() })
    }

    pub fn cursor(&self) -> u64 { self.base_cursor + self.ops.len() as u64 }
    pub fn operations(&self) -> &[Value] { &self.ops }

    pub fn submit(&mut self, operations: &[Value]) -> Result<Vec<Value>> {
        if self.max_batch_ops.is_some_and(|limit| operations.len() > limit) { return Err(PolypackError::ResourceLimit { name: "maxBatchOps".into(), limit: self.max_batch_ops.unwrap() }); }
        // `validate_sync_batch` also computes a whole-batch checksum (a full
        // JSON serialize + UTF-16 hash pass) that submit() never uses —
        // validate each operation directly instead of paying for a checksum
        // that's immediately discarded.
        for operation in operations { validate_sync_operation(operation)?; }
        let mut accepted = Vec::new();
        let mut accepted_transactions = HashSet::new();
        for operation in operations {
            let object = operation.as_object().unwrap();
            let client_id = object["clientId"].as_str().unwrap();
            let seq_key = format!("{client_id}:{}", object["seq"].as_u64().unwrap());
            let operation_key = object.get("operationId").and_then(Value::as_str).map(|id| format!("{client_id}:{id}"));
            let transaction_key = object.get("transactionId").and_then(Value::as_str).map(|id| format!("{client_id}:{id}"));
            if self.seen_ops.contains(&seq_key)
                || operation_key.as_ref().is_some_and(|key| self.operation_ids.contains(key))
                || transaction_key.as_ref().is_some_and(|key| self.transaction_ids.contains(key) && !accepted_transactions.contains(key))
            { continue; }
            self.seen_ops.insert(seq_key);
            if let Some(key) = &operation_key { self.operation_ids.insert(key.clone()); }
            if let Some(key) = &transaction_key { self.transaction_ids.insert(key.clone()); accepted_transactions.insert(key.clone()); }
            // One clone stays in `self.ops`, the other is returned to the
            // caller — both are needed, but doing them together here (rather
            // than cloning into `accepted` and then cloning `accepted` again
            // into `self.ops` afterwards) avoids a second full pass over the
            // accepted operations.
            self.ops.push(operation.clone());
            accepted.push(operation.clone());
        }
        if let Some(limit) = self.max_ops {
            if self.ops.len() > limit {
                let removed = self.ops.len() - limit;
                // Only the seq-keyed entry is bounded by the ring buffer;
                // operationId/transactionId identities are retained forever
                // so a delayed retry after compaction is still deduped
                // instead of being re-accepted. Matches the TS/Python servers.
                for op in self.ops.drain(..removed) {
                    let object = op.as_object().unwrap();
                    let key = format!("{}:{}", object["clientId"].as_str().unwrap(), object["seq"].as_u64().unwrap());
                    self.seen_ops.remove(&key);
                }
                self.base_cursor += removed as u64;
            }
        }
        Ok(accepted)
    }

    pub fn recover(&self, from_cursor: u64, limit: usize) -> Result<Value> {
        let valid = from_cursor >= self.base_cursor && from_cursor <= self.cursor();
        let requested = if valid { from_cursor } else { 0 };
        let offset = if requested == 0 { 0 } else { (requested - self.base_cursor) as usize };
        let page: Vec<Value> = self.ops.iter().skip(offset).take(limit).cloned().collect();
        let cursor = (if requested == 0 { self.base_cursor } else { requested }) + page.len() as u64;
        let checksum = sync_checksum(&page);
        let mut response = serde_json::json!({ "type": if requested == 0 { "snapshot" } else { "delta" }, "clientId": "server", "fromSeq": requested, "cursor": cursor, "more": offset + page.len() < self.ops.len(), "ops": page, "protocolVersion": self.protocol_version, "checksum": checksum });
        if !valid { response["errors"] = serde_json::json!([{ "code": "cursor_expired", "message": "requested cursor is no longer available" }]); }
        Ok(response)
    }
}
